README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# minizdog <img src="man/figures/logo.png" align="right" height=230/>

<!-- badges: start -->

![](http://img.shields.io/badge/cool-useless-green.svg)
![](http://img.shields.io/badge/mini-verse-blue.svg) [![Travis build
status](https://travis-ci.org/coolbutuseless/minizdog.svg?branch=master)](https://travis-ci.org/coolbutuseless/minizdog)
<!-- badges: end -->

`minizdog` is a package for building [Zdog](https://zzz.dog)
illustrations in R.

This package was inspired by [OganM](https://twitter.com/OganM)’s
package called [rdog](https://github.com/oganm/rdog). That package is
much more feature complete, with better documentation and examples and
other stuff.

This package is part of an ongoing exploration of writing document
interfaces with R and [R6](https://cran.r-project.org/package=R6) -
every drawing feature is mapped to an R6 object, and these objects are
nested within each other to create the document.

## Installation

You can install the development version from
[GitHub](https://github.com/coolbutuseless/minizdog) with:

``` r
# install.packages("devtools")
devtools::install_github("coolbutuseless/minizdog")
```

## All-in-one example

``` r
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Initialize a document
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
zdog <- ZdogIllustration$new(zoom = 2, yincrement = 0.03)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Create a circle as a separate object
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ell <- minizdog::ZdogEllipse$new(diameter = 20, stroke = 5, color = '#636')$
  translate(z = -20)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Create a rectangle directly as a child of the illustration
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
zdog$rect(width = 20, height = 20, stroke = 3)$
  update(fill = TRUE, color = '#e62')$
  translate(z = -10)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Create a polygon using the `ztag` helper
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
poly <- ztag$polygon(radius = 40, sides = 6)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# create some text
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
text <- ZdogText$new(value = "#RStats\\n\\nmini\\nzdog")$
  update(color = '#00f', textAlign = 'center', fontSize = 20)$
  translate(y = 15)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Append the standalone elements to the main document
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
zdog$append(ell)
zdog$append(poly)
zdog$append(text)
```

``` r
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# If running in Rstudio this will open the zdog illustration in the viewer pane.
# Github limits what a README can show, so this is just a GIF animation of
# what the above javascript creates
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
zdog$show()
```

<img src = "man/figures/badlogo-o.gif">

``` r
as.character(zdog) # will print just the javascript code
zdog$as_html()     # wraps the javascript in some html boilerplate
```

<details close>

<summary> <span title="as.character(zdog) output (JS)">
as.character(zdog) output (JS)</span> </summary>

    let isSpinning = true;
    Zfont.init(Zdog);
    var default_font = new Zdog.Font({
      src: 'https://cdn.jsdelivr.net/gh/jaames/zfont/demo/fredokaone.ttf'
    });
          
    let Ill01 = new Zdog.Illustration({
      element: '.zdog-canvas',
      zoom: 2,
      centered: true,
      dragRotate: true,
      onDragStart: function() {isSpinning = false;}
    });
      let Rect01 = new Zdog.Rect({
        stroke: 3,
        width: 20,
        height: 20,
        addTo: Ill01,
        fill: true,
        color: '#e62',
        translate: { z: -10 }
      });
      let Ellipse01 = new Zdog.Ellipse({
        stroke: 5,
        color: '#636',
        diameter: 20,
        translate: { z: -20 },
        addTo: Ill01
      });
      let Polygon01 = new Zdog.Polygon({
        radius: 40,
        sides: 6,
        addTo: Ill01
      });
      let Text01 = new Zdog.Text({
        value: '#RStats\n\nmini\nzdog',
        fontSize: 20,
        textAlign: 'center',
        textBaseline: 'left',
        font: default_font,
        color: '#00f',
        translate: { y: 15 },
        addTo: Ill01
      });
    function animate() {
      // rotate illo each frame
      if (isSpinning) {
        Ill01.rotate.x += 0;
        Ill01.rotate.y += 0.03;
        Ill01.rotate.z += 0;
      }
      Ill01.updateRenderGraph();
      // animate next frame
      requestAnimationFrame( animate );
    }
    // start animation
    animate();

</details>

<br />

<details open>

<summary> <span title="zdog$as_html() output"> zdog$as.html() output
</span> </summary>

    <!doctype html>
    <html lang='en'>
    <head>
    <meta charset='UTF-8'>
    </head>
    <body>
    <canvas id='canvas' class='zdog-canvas' width = '300', height = '300'>
    Your viewer doesn't support HTML5 canvas
    </canvas>
    
    <script src='https://unpkg.com/zdog@1/dist/zdog.dist.min.js'></script>
    <script src='https://cdn.jsdelivr.net/npm/zfont/dist/zfont.min.js'></script>
    <script>
    let isSpinning = true;
    Zfont.init(Zdog);
    var default_font = new Zdog.Font({
      src: 'https://cdn.jsdelivr.net/gh/jaames/zfont/demo/fredokaone.ttf'
    });
          
    let Ill01 = new Zdog.Illustration({
      element: '.zdog-canvas',
      zoom: 2,
      centered: true,
      dragRotate: true,
      onDragStart: function() {isSpinning = false;}
    });
      let Rect01 = new Zdog.Rect({
        stroke: 3,
        width: 20,
        height: 20,
        addTo: Ill01,
        fill: true,
        color: '#e62',
        translate: { z: -10 }
      });
      let Ellipse01 = new Zdog.Ellipse({
        stroke: 5,
        color: '#636',
        diameter: 20,
        translate: { z: -20 },
        addTo: Ill01
      });
      let Polygon01 = new Zdog.Polygon({
        radius: 40,
        sides: 6,
        addTo: Ill01
      });
      let Text01 = new Zdog.Text({
        value: '#RStats\n\nmini\nzdog',
        fontSize: 20,
        textAlign: 'center',
        textBaseline: 'left',
        font: default_font,
        color: '#00f',
        translate: { y: 15 },
        addTo: Ill01
      });
    function animate() {
      // rotate illo each frame
      if (isSpinning) {
        Ill01.rotate.x += 0;
        Ill01.rotate.y += 0.03;
        Ill01.rotate.z += 0;
      }
      Ill01.updateRenderGraph();
      // animate next frame
      requestAnimationFrame( animate );
    }
    // start animation
    animate();
    </script>
    
    </body>
    </html>

</details>

<br />


NEWS.md
# minizdog 0.1.0

* Initial release


R/ZdogText.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' @rdname ZdogShapeClass
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ZdogText <- R6::R6Class(
  "ZdogText", inherit = ZdogShapeClass,

  public = list(
    initialize = function(..., value, font = 'default_font', fontSize = 64,
                          textAlign = c('left', 'center', 'right'),
                          textBaseline = c('bottom', 'top', 'middle')) {

      textAlign    <- match.arg(textAlign)
      textBaseline <- match.arg(textAlign)

      super$initialize(...,
                       type         ='Text',
                       value        = value,
                       fontSize     = fontSize,
                       textAlign    = textAlign,
                       textBaseline = textBaseline,
                       font         = font)

      self
    }
  )
)


R/ZdogShape.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Zdog element builder
#'
#' @import R6
#' @import glue
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ZdogShapeClass <- R6::R6Class(
  "ZdogShapeClass", inherit = ZdogAnchor,

  public = list(

    initialize = function(..., type = 'Shape') {
      super$initialize(type=type, ...)
    },


    color = function(col) {
      self$attribs$color <- paste0("'", col, "'")
      invisible(self)
    },

    stroke = function(width) {
      if (is.logical(width)) {
        self$attribs$stroke <- tolower(as.character(width))
      } else {
        self$attribs$stroke <- width
      }
      invisible(self)
    },

    fill = function(fill_lgl) {
      self$attribs$fill <- tolower(as.character(isTRUE(fill_lgl)))
      invisible(self)
    },

    closed = function(closed_lgl) {
      self$attribs$closed <- tolower(as.character(isTRUE(closed_lgl)))
      invisible(self)
    },

    visible = function(visible_lgl) {
      self$attribs$visible <- tolower(as.character(isTRUE(visible_lgl)))
      invisible(self)
    },

    backface = function(backface_lgl) {
      self$attribs$backface <- tolower(as.character(isTRUE(backface_lgl)))
      invisible(self)
    },

    front = function(...) {
      stop("front not done yet. need generic vector collapse")
    }
)
)



if (FALSE) {
  ZdogShapeClass$new(id = "greg")
}

R/ZdogAnchor.R


collapse_vector <- function(x = NULL, y = NULL, z = NULL) {
  vec <- c(x = x, y = y, z = z)
  vec <- Filter(Negate(is.null), vec)
  res <- paste(names(vec), unlist(vec), sep = ": ", collapse = ", ")
  res <- paste("{", res, "}")
  res
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Generate some nicer looking ID strings
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
IDGenerator <- R6::R6Class(
  "IDGenerator",

  public = list(
    count = NULL,

    initialize = function() {
      self$count <- list()
    },

    next_id = function(prefix = 'zdog') {
      if (prefix %in% names(self$count)) {
        self$count[[prefix]] <- self$count[[prefix]] + 1L
      } else {
        self$count[[prefix]] <- 1L
      }
      sprintf("%s%02i", prefix, self$count[[prefix]])
    }
  )
)


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Zdog element builder
#'
#' @import R6
#' @import glue
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ZdogAnchor <- R6::R6Class(
  "ZdogAnchor",

  public = list(

    id_gen   = IDGenerator$new(),
    id       = NULL,
    type     = NULL,
    attribs  = NULL,
    children = NULL,

    initialize = function(..., type = 'Anchor') {
      self$type     <- type
      self$update_id()

      self$attribs  <- list()
      self$children <- list()

      self$update(...)

      self
    },

    update_id = function(recursive = FALSE) {
      self$id <- self$id_gen$next_id(self$type)

      if (recursive) {
        for (child in self$children) {
          child$attribs$addTo <- self$id
          child$update_id(recursive = TRUE)
        }
      }
      invisible(self)
    },


    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Update the HTML Element.
    #   - Named arguments are considered attributes and will overwrite
    #     existing attributes with the same name. Set to NULL to delete the attribute
    #   - Unnamed arguments are appended to the list of child nodes.  These
    #     should be text, other ZdogAnchors or any ojbect that can be represented
    #     as a single text string using "as.character()"
    #   - to print just the attribute name, but without a value, set to NA
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    update = function(...) {
      varargs      <- list(...)
      if (length(varargs) == 0) {
        return(invisible(self))
      }

      vararg_names <- names(varargs)
      if (is.null(vararg_names)) {
        vararg_names <- character(length = length(varargs))
      }
      has_name   <- nzchar(vararg_names)

      children <- varargs[!has_name]
      attribs  <- varargs[ has_name]

      self$attribs  <- modifyList(self$attribs, attribs, keep.null = FALSE)
      do.call(self$append, children)

      invisible(self)
    },


    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Insert a child node. by default at the end of the list of children nodes
    # but 'position' argument can be used to set location
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    append = function(..., position = NULL) {
      child_objects <- list(...)

      for (x in child_objects) {
        x$update(addTo = self$id)
      }

      if (is.null(position)) {
        self$children <- append(self$children, child_objects)
      } else {
        self$children <- append(self$children, child_objects, after = position - 1)
      }

      invisible(self)
    },

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # create an HTML element and add it to the document. return the newly
    # created element
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    add = function(name, ...) {
      if (!is.character(name)) {
        stop("ZdogAnchor$add(): 'name' must be a character string")
      }
      new_elem <- ZdogAnchor$new(name, ...)
      self$append(new_elem)
      invisible(new_elem)
    },


    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Remove child objects at the given indicies
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    remove_indices = function(indicies) {
      self$children[indices] <- NULL
      inviisible(self)
    },


    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Recursively convert this XMLElement and children to text and concatenate
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    as_character_inner = function(..., depth = 0) {
      indent1   <- create_indent(depth)
      indent2   <- create_indent(depth + 1)

      value_attribs <- self$attribs

      # Convert all to character
      value_attribs <- lapply(value_attribs, function(x) {
        if (is.logical(x)) {
          tolower(as.character(isTRUE(x)))
        } else if (inherits(x, 'ZdogAnchor')) {
          x$id
        } else {
          x
        }
      })

      # names of colours should be quoted
      quoted_attributes <- c('element', 'color', 'value', 'textAlign', 'textBaseline')
      for (attr_name in quoted_attributes) {
        if (attr_name %in% names(value_attribs)) {
          value_attribs[[attr_name]] <- sQuote(value_attribs[[attr_name]], FALSE)
        }
      }

      if (length(value_attribs) > 0) {
        attr_names <- paste0(indent2, names(value_attribs))
        value_attribs <- paste(attr_names,
                               unlist(value_attribs),
                               sep = ": ", collapse = ",\n")
        # value_attribs <- paste0(value_attribs, ",")
      } else {
        value_attribs <- NULL
      }

      attribs <- value_attribs
      open    <- glue::glue("{indent1}let {self$id} = new Zdog.{self$type}({{")
      close   <- glue::glue("{indent1}}});")

      if (length(self$children) > 0) {
        children <- lapply(self$children, as.character, depth = depth + 1)
        children <- unlist(children, use.names = FALSE)
      } else {
        children = NULL
      }

      c(open, value_attribs, close, children)
    },


    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Recursively convert this XMLElement and children to text and concatenate
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    as_character = function(..., depth = 0) {
      zdog_string <- paste0(self$as_character_inner(depth = depth), collapse = "\n")

      zdog_string
    },


    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Print the HTML string
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    print = function(...) {
      cat(self$as_character(...))
      invisible(self)
    },

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Save HTML fragment
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    save = function(filename, ...) {
      writeLines(self$as_character(...), filename)
      invisible(self)
    },


    addTo = function(x) {
      if (inherits(x, "ZdogAnchor")) {
        self$attribs$addTo <- x$id
      } else {
        self$attribs$addTo <- as.character(x)
      }
      invisible(self)
    },

    translate = function(x=NULL, y=NULL, z=NULL) {
      self$attribs$translate <- collapse_vector(x = x, y = y, z = z)
      invisible(self)
    },

    rotate = function(x = NULL, y = NULL, z = NULL) {
      if (!is.null(x)) {
        self$attribs$rotate <- collapse_vector(x = x)
      } else if (!is.null(y)) {
        self$attribs$rotate <- collapse_vector(y = y)
      } else if (!is.null(z)) {
        self$attribs$rotate <- collapse_vector(z = z)
      }
      invisible(self)
    },

    scale = function(x, y = NULL, z = NULL) {
      self$attribs$scale <- collapse_vector(x = x, y = y, z = z)
      invisible(self)
    },

    addChild = function(x) {
      self$append(x)
      invisible(self)
    },

    removeChild = function(x) {
      warning("removeChild() does not currently have any effect")
      invisible(self)
    },

    remove = function() {
      warning("remove() does not currently have any effect")
      invisible(self)
    },

    updateGraph = function() {
      warning("updateGraph() does not currently have any effect")
      invisible(self)
    },

    renderGraphCanvas = function() {
      warning("renderGraphCanvas() does not currently have any effect")
      invisible(self)
    },

    renderGraphSvg = function(...) {
      warning("renderGraphSvg() does not currently have any effect")
      invisible(self)
    },

    normalizeRotate = function() {
      warning("normalizeRotate() does not currently have any effect")
      invisible(self)
    },

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Shallow copy just this element. and empty out the children list.
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    copy = function() {
      new <- self$clone(deep = FALSE)
      new$update_id()
      new$children <- list()
      new
    },

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Deep copy needed as 'children' is a list of R6 objects.
    # update this ID and cascade an ID/addTo update across all children
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    copyGraph = function() {
      new <- self$clone(deep = TRUE)
      new$update_id(recursive = TRUE)
      new
    }
  ),


  private = list(
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # When called with `$clone(deep = TRUE)`, the 'deep_clone' function is
    # called for every name/value pair in the object.
    # See: https://r6.r-lib.org/articles/Introduction.html
    # Need special handling for:
    #   - 'children' is a list of R6 objects
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    deep_clone = function(name, value) {
      if (name %in% c('children')) {
        lapply(value, function(x) {if (inherits(x, "R6")) x$clone(deep = TRUE) else x})
      } else {
        value
      }
    }
  )

)


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Retrieve character representation of ZdogAnchor
#'
#' @param x ZdogAnchor object
#' @param ... other arguments
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
as.character.ZdogAnchor <- function(x, ...) {
  x$as_character(...)
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' @rdname ZdogAnchor
#' @usage NULL
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
zdog_anchor <- function(...) {
  ZdogAnchor$new(...)
}




if (FALSE) {
  rect1 <- ztag$rect(width = 64, height = 64, stroke = 16, color = '#ea0')$
    translate(x = -40)

  rect2 <- rect1$copy()$
    translate(x = 48)$
    update(color = '#c25')

  zdog <- zdog_ill(rect1, rect2)
  zdog
  rect1
}


if (FALSE) {
  rect1 <- ztag$rect(width = 64, height = 64, stroke = 16, color = '#ea0')$
    translate(x = -40)

  ss <- rect1$shape(x = 0, y=0, color = '#00f', stroke = 10)
  ss

  rect2 <- rect1$copyGraph()$
    translate(x = 48)$
    update(color = '#c25')

  zdog <- zdog_ill(rect1, rect2)
  zdog
  rect1

  zdog$show()
}






















R/ZdogIllustration.R

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' HTML5 Zdog Builder
#'
#' @import R6
#' @import glue
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ZdogIllustration <- R6::R6Class(
  "ZdogIllustration", inherit = ZdogAnchor,

  public = list(
    width      = NULL,
    height     = NULL,
    xincrement = NULL,
    yincrement = NULL,
    zincrement = NULL,
    font_name  = NULL,
    font_src   = NULL,

    initialize = function(..., width = 300, height = 300,
                          element    = '.zdog-canvas',
                          zoom       = NULL,
                          centered   = TRUE,
                          dragRotate = TRUE,
                          resize     = NULL,
                          yincrement = 0.01,
                          xincrement = 0,
                          zincrement = 0
                          ) {


      self$width      <- width
      self$height     <- height
      self$xincrement <- xincrement
      self$yincrement <- yincrement
      self$zincrement <- zincrement
      self$font_name  <- list()
      self$font_src   <- list()

      super$initialize(
        ...,
        type       ='Illustration',
        element    = element,
        zoom       = zoom,
        centered   = centered,
        dragRotate = dragRotate,
        resize     = resize,
        onDragStart = "function() {isSpinning = false;}"
      )

      self
    },


    add_font = function(font_name, font_src) {
      self$font_name <- c(self$font_name, font_name)
      self$font_src  <- c(self$font_src , font_src )
      invisible(self)
    },


    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Character representatino of just the javascript for this zdog document
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    as_character = function(...) {
      open <- "let isSpinning = true;
Zfont.init(Zdog);
var default_font = new Zdog.Font({
  src: 'https://cdn.jsdelivr.net/gh/jaames/zfont/demo/fredokaone.ttf'
});
      "

      fonts <- glue("var {self$font_name} = new Zdog.Font({{ src: '{self$font_src}' }});")

      res <- super$as_character(...)

      close <- glue("
// update & render
{self$id}.updateRenderGraph();
")

      close <- glue("
function animate() {{
  // rotate illo each frame
  if (isSpinning) {{
    {self$id}.rotate.x += {self$xincrement};
    {self$id}.rotate.y += {self$yincrement};
    {self$id}.rotate.z += {self$zincrement};
  }}
  {self$id}.updateRenderGraph();
  // animate next frame
  requestAnimationFrame( animate );
}}
// start animation
animate();")


      paste(c(open, fonts, res, close), collapse = "\n")
    },


    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Character representation of the full HTML + Javascript to render this
    # ZdogIllustration
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    as_html = function(...) {
      html_string <- glue("<!doctype html>
<html lang='en'>
<head>
<meta charset='UTF-8'>
</head>
<body>
<canvas id='canvas' class='zdog-canvas' width = '{self$width}', height = '{self$height}'>
Your viewer doesn't support HTML5 canvas
</canvas>

<script src='https://unpkg.com/zdog@1/dist/zdog.dist.min.js'></script>
<script src='https://cdn.jsdelivr.net/npm/zfont/dist/zfont.min.js'></script>
<script>
{self$as_character()}
</script>

</body>
</html>")
      html_string
    },

    zoom = function(zoom) {
      self$attribs$zoom <- zoom
      invisible(self)
    },

    centered = function(centered) {
      self$attribs$centered <- centered
      invisible(self)
    },

    dragRotate = function(dragRotate) {
      self$attribs$dragRotate <- dragRotate
      invisible(self)
    },

    resize = function(resize) {
      self$attribs$resize <- resize
      invisible(self)
    },


    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Output zdog document to screen
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    print = function(...) {
      cat(self$as_character())
    },


    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Save HTML + JS to fully render this ZdogIllustration
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    save_html = function(filename) {
      writeLines(self$as_html(), filename)
    },


    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # View HTML in whatever viewer is available
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    show = function(viewer = getOption("viewer", utils::browseURL)) {
      www_dir <- tempfile("viewhtml")
      dir.create(www_dir)
      index_html <- file.path(www_dir, "index.html")
      self$save_html(index_html)

      if (!is.null(viewer)) {
        viewer(index_html)
      } else {
        warning("No viewer available.")
      }
      invisible(index_html)
    }
  )
)


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Helper to create ZdogIllustration objects
#'
#' @param width,height Zdog dimensions
#' @param ... Initialise the zdog doc with these elements
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
zdog_ill <- function(..., width = 240, height = 240) {
  ZdogIllustration$new(..., width = width, height = height)
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert a ConvasDocument to a string
#'
#' @param x ZdogIllustration object
#' @param ... other arguments
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
as.character.ZdogIllustration <- function(x, ...) {
  x$as_character(...)
}





if (FALSE) {
  zdog <- zdog_ill(zoom = 2)

  ell <- ZdogEllipse$new(diameter = 20, stroke = 5, color = '#636')$
    translate(z = 10)

  zdog$rect(width = 20, height = 20, stroke = 3)$
    update(fill = TRUE, color = '#e62')$
    translate(z = -10)

  text <- ZdogText$new(value = "Crap", color = '#00f', textAlign = 'center')

  zdog$append(ell)
  # zdog$append(rect)
  zdog$append(text)
  zdog

  zdog$show()

}

